GSS 2.0 documentation

cbchouinard opened the pull request:
I think at this stage we have enough docs to get people up and running with GSS. 

I'll had more docs in the next weeks for more advance usage scenarios. I'll also add some more complete examples and tutorials. 

I would like to have your comments before publishing. I'll fix the merging conflicts tomorrow.

Thanks!

paulyoung commented on guides/_posts/2014-01-11-ifelse.html:
```diff
@@ -36,73 +49,83 @@
+      right: == :next[left];
+    }
+  }
 }
```
I think this could be simplified a lot to remove other concepts and put the most basic point across.

Something like this:

```
section {
  article {
    @if ^[width] <= 400 {
      width: == ^[width];
      height: == ^[height] / 2;
    }
    @else  {
      width: == ^[width] / 2;
      height: == ^[height];
    }
  }
}
```

paulyoung commented on guides/_posts/2014-01-11-ifelse.html:
```diff
@@ -36,73 +49,83 @@
 {% endhighlight %}
 
+When the `section` element `width` goes bellows 400px, the `article` elements will be constrain
+by the `@if` block otherwise by the `@else` block.
```
There are a couple of typos here, but I also think this should read:

> When the width of the `section` element goes below 400px, the `article` elements will be constrained by the `@if` block, otherwise by the `@else` block.

paulyoung commented on guides/_posts/2014-01-11-ifelse.html:
```diff
@@ -36,73 +49,83 @@
+When the `section` element `width` goes bellows 400px, the `article` elements will be constrain
+by the `@if` block otherwise by the `@else` block.
+
+### Constraints evaluation in conditions
```
Suggesting that we change this to "Constraint evaluation within conditionals"

paulyoung commented on guides/_posts/2014-01-11-ifelse.html:
```diff
@@ -36,73 +49,83 @@
+
+### Constraints evaluation in conditions
+In order to determine if some conditions hold, GSS needs to evaluate every constraints outside of those conditions before evaluating if
+they are met, which include constraints appearing after the conditions declaration.
```
Some mixing of plural and singular here.

paulyoung replied:
I find it easier to think about if it's phrased like this:

"In order to determine if a condition holds, GSS needs to evaluate every constraint outside of that conditional before evaluating if it is met. This includes constraints appearing after the conditional declaration."

paulyoung commented on guides/_posts/2014-01-11-ifelse.html:
```diff
@@ -36,73 +49,83 @@
+they are met, which include constraints appearing after the conditions declaration.
+
+As explained in the [GSS constraints guide](/guide/constraints), the order of constraints declaration can affects the
+resulting solution.
```
I believe these should be singular: "the order of _constraint_ declaration can _affect_ the resulting solution."

paulyoung commented on guides/_posts/2014-01-11-ifelse.html:
```diff
@@ -36,73 +49,83 @@
+resulting solution.
 
+The constraint within the condition of the following example will be added to the solver after the outside constraints. Which
+make sens since the condition is true, you want the constraint within that condition to overcome the other evenly strenghted ones.
```
Suggesting we reword it as:

> In the following example, the constraint within the conditional will be added to the solver after the constraints outside of it. This causes the constraint within the conditional to overcome the other constraints of equal strength when the conditional evaluates as true.

This addresses some typos, singular vs plural terms, and removes the use of the pronoun "you" which is less appropriate in formal writing.

paulyoung commented on guides/_posts/2014-01-11-ifelse.html:
```diff
@@ -36,73 +49,83 @@
-{% endhighlight %}
 
+#elm[width] == 300;
+#elm[height] == 500;
```
The use of 2 width constraints here is redundant in relation to the point about conditionals. Can we remove one of them to keep things simple?

paulyoung commented on guides/_posts/2014-01-11-ifelse.html:
```diff
@@ -36,73 +49,83 @@
-{% highlight css %}
-#box1[width] == 9;
-#box2[width] == 19;
+In this example, if the `width` of the screen is smaller than 400px, the `height` of the element will be constrain to 300px even
```
Should be "will be _constrained_ to 300px"

paulyoung commented on guides/_posts/2014-01-11-ifelse.html:
```diff
@@ -36,73 +49,83 @@
-#box1[width] == 9;
-#box2[width] == 19;
+In this example, if the `width` of the screen is smaller than 400px, the `height` of the element will be constrain to 300px even
+though there's a `height` constraint defined after the condition.
```
I think in most cases we mean "conditional" (statement) and not "condition".

paulyoung commented on guides/_posts/2014-01-11-ifelse.html:
```diff
@@ -36,73 +49,83 @@
-}
-{% endhighlight %}
+### Condition looping
+As describe in the [GSS constraints guide](/guide/constraints), constraints are two ways. You need to be well aware of this fact
```
As _described_ in the

paulyoung replied:
constraints are _two-way_.

paulyoung commented on guides/_posts/2014-01-11-ifelse.html:
```diff
@@ -36,73 +49,83 @@
+As describe in the [GSS constraints guide](/guide/constraints), constraints are two ways. You need to be well aware of this fact
+when using conditions. GSS is pretty good at handling cases where a looping could occurs but still, be aware that in some
+complexe conditions nesting you could have some situation where you get in a loop but GSS should get your back the vast majority
+of the time.
```
I'd prefer it if we were less informal, so instead of this:

> You need to be well aware of this fact when using conditions.

could we say something like this?

> This should be carefully considered when using conditionals.

paulyoung replied:
Similar thing with:

> GSS is pretty good at handling cases where a looping could occurs but still, be aware that in some complexe conditions nesting you could have some situation where you get in a loop but GSS should get your back the vast majority of the time.

Can we say something like this?

> Although GSS has been designed to detect and prevent cases when looping could occur, complex nesting of conditionals can still cause infinite loops.

(also addresses some typos and mixed tense)

paulyoung commented on guides/_posts/2014-01-11-ifelse.html:
```diff
@@ -36,73 +49,83 @@
+of the time.
 
+For example, the following could have led to a loop, but the GSS cassowary solver will figure out the solution for this set of
+constraints without looping:
```
Here we could explain why it would have lead to a loop.

paulyoung replied:
Is this actually cassowary preventing the loop or some special handling?

paulyoung replied:
@Inviz is this accurate?

paulyoung replied:
Suggesting we rewrite this as:

> For example, since the conditional below changes a variable that is being checked, it would appear that it will lead to an infinite loop. However, the underlying constraint solver in GSS can determine a solution that satisfies constraints without looping.

(this assumes that this is even true)

paulyoung commented on guides/_posts/2014-01-11-ifelse.html:
```diff
@@ -36,73 +49,83 @@
-WIP, more coming soon...
+## Read next
+With only CCSS constraints at your disposal, constraining common layout scenarios quickly becomes tedious. Read our
+[VFL guide](/guides/ifelse) to learn how to more efficiently constraint your layout.
```
Should link to `/guides/vfl`, not `/guides/ifelse`.

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -22,26 +34,26 @@
 /* which is the same as */
 
-@horizontal |-10-[#button1]-10-[#button2]-10-[#button3]-10-[#button4]-10-| in(#panel);
+@horizontal |-10-(#button1)-10-(#button2)-10-(#button3)-10-(#button4)-10-| in(#panel);
```
I personally think we should promote the use of `@h` and `@v` exclusively and remove other ways from the CCSS parser. Would be good to get some other people's input.

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -22,26 +34,26 @@
-Horizontal layouts can be created with the CSS directive `@h`, `@horizontal` or `@-gss-horizontal`.  
+Horizontal layouts can be created with the CSS directive `@h`, `@horizontal` or `@-gss-horizontal`.
 
 Vertical layouts with the CSS directive `@v`, `@vertical` or `@-gss-vertical`.
```
Relevant to the previous comment.

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -88,29 +100,29 @@
 
```
I think this is still true, but a default value is no longer defined (I believe it used to be 8).

@d4tocchini and I had discussed throwing an error when using a hyphen without the gap function.

I think we should get finalize our position on this.

paulyoung replied:
In case it's hidden, it's the line that reads "If no `gap()` is defined, but a hyphen is used to make a connection then a `standard gap` variable is created,"

paulyoung replied:
That line should also end in a period and not a comma.

paulyoung commented on guides/_posts/2014-01-20-ccss.html:
```diff
@@ -136,14 +136,14 @@
 
 /* the !require strength ensures this constraint will hold */
-[col-size] == ::window[width] / 12 !require; 
+[col-size] == ::window[width] / 12 !require;
```
Use of square brackets here seems inconsistent with other examples I've seen that don't have them.

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -121,36 +133,37 @@
-  ::[right] - 50 == #purple-box[right];
+  &[left] + 50 == #purple-box[left];
+  &[right] - 50 == #purple-box[right];
 }
```
This is simpler and consistent with the previous example:

```
#container[left] + 50 == #purple-box[left];
#container[right] - 50 == #purple-box[right];
```

paulyoung replied:
Ah, I see. We're specifically highlighting the use of `&` as `::this`.

paulyoung replied:
Might be worth mentioning that it's not necessary in this case, since we can just do this:

```
#container {
  left: == #purple-box[left] - 50;
  right: == #purple-box[right] + 50;
 }
```

Not sure if we can omit the colon here.

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -88,29 +100,29 @@
 {% endhighlight %}
 
-The `[hgap]` & `[vgap]` variables can be used like normal variables, and again like all constraint variables are globally scoped.
+The `hgap` & `vgap` variables can be used like normal variables, and again like all constraint variables are globally scoped.
```
As above. I don't think we define these as global variables anymore.

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -219,31 +233,15 @@
+## Cushions
 
-Cushions are similar to connections, only they use inequality instead of equality constraints. 
+Cushions are similar to connections gaps, only they use inequality instead of equality constraints.
```
similar to _connection_ gaps,

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -219,31 +233,15 @@
 {% highlight css %}
 /* VFL */
-@horizontal [#box1]~[#box2];
+@h (#box1)~(#box2);
```
I'm not sure what this does. What is `~` when used as a gap? Is this the same as `~-~`?

paulyoung replied:
Same concerns about lack of `gap()` function.

cbchouinard replied:
We show the equivalent CCSS underneath it: if #box1[right] <= #box2[left]

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -259,8 +257,8 @@
-@horizontal [#box1]~-~[#box2];
-@vertical [#box3]~-~[#box4];
+@h (#box1)~-~(#box2);
+@v (#box3)~-~(#box4);
```
Mentioned this before, but I don't think we should encourage using gaps like this without using `gap()`.

cbchouinard replied:
In this case an implicit gap hgap or vgap will be use and bind to the current scope, I agree that hoisting could lead to ambiguity for the user. I'll add a note in the doc.

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -300,34 +301,144 @@
 {% highlight css %}
 /* VFL */
-@horizontal [#flexibleButton(>= 70, <= 100))];
+@h (#flexibleButton(>= 70, <= 100)));
```
Parentheses are unbalanced here.

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -300,34 +301,144 @@
+
+## VFL Selectors
+All previous example of VFL expressions use basic [selectors](/guides/selectors). But you can also use combinators in VFL expression
+like this:
```
All previous _examples_ of

paulyoung replied:
Could we reword as this?

> Although all previous examples of VFL expressions only use basic selectors, combinators can also be used in VFL expressions like so:

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -300,34 +301,144 @@
 {% highlight css %}
-/* VFL */
-@horizontal |-24-[#find]-[#findNext]-[#findField(>=30)]-24-| in(#popup);
+@v | (.featured > article ~ .title) | in(#elm > .class .class2"_fallout")
```
What is `.class2"_fallout"`? Is that a typo?

paulyoung replied:
In any case, I propose that we call the virtual `"virtual"` or something less obscure.

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -300,34 +301,144 @@
+<a name="vflsplats"></a>
+## VFL Splat
+When a selector in your VFL expression finds more than one element, due to [plural binding](/guides/selectors/#pluralbinding),
+if the number of element within each selectors expression are not equal, then some elements will not be constraint.
```
I feel like something like this is easier to understand:

> When a selector in a VFL expression finds more than one element, not all elements will be constrained if the number of elements within each selector's expression are not equal. This is due to plural binding.

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -300,34 +301,144 @@
+<a href="http://codepen.io/cbchouinard/pen/vEJqbG" target="_blank">Live example.</a>
+
+In this example, the last two `article` elements will not be constrained since there's no `section` element to bind to.
+As you can see in the CCSS generated constraints, the elements are pluraly binded.
```
Suggestion:

> In this example, the last two `article` elements will not be constrained since there is no `section` element to bind to. As can be seen in the generated CCSS, the elements are constrained using plural binding.

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -300,34 +301,144 @@
+In this example, the last two `article` elements will not be constrained since there's no `section` element to bind to.
+As you can see in the CCSS generated constraints, the elements are pluraly binded.
+
+{% highlight css %}
```
Might want to make this `{% highlight html %}`

paulyoung replied:
In fact, yeah, it's all HTML.

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -300,34 +301,144 @@
+{% endhighlight %}
+
+If the number of selected elements matches, the position of the elements within each collection of selected
+elements will overlap since you're basically constraining the whole collection to the same values. <a href="http://codepen.io/cbchouinard/pen/yyodWx" target="_blank">Live example.</a>
```
I think this needs some work.

paulyoung replied:
The linked codepen has some elements that have `class="notBinded"`, which I think contradicts the point we're trying to make here since the number of elements is equal.

Also, I think the preferred past tense verb for "bind" should be "bound".

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -300,34 +301,144 @@
+/* the 3 sections will overlap over each other as the 3 articles. */
+@h |-(section)-(article)-(footer)-| in(::window);
+
+</style>
```
Related to the above wording, I think rather than just say that the elements "overlap" we could say that the elements all have the same _horizontal position_.

Since we're not specifying anything about their vertical position, they don't necessarily overlap.

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -300,34 +301,144 @@
+Splats will iterate over the collection of selected elements and position elements
+next to each other. <a href="http://codepen.io/cbchouinard/pen/ogerrX" target="_blank">Live example.</a>
+
+{% highlight css %}
```
css -> html

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -300,34 +301,144 @@
+
+<style type="text/gss">
+
+/* elements are not 'pluraly' binded */
```
Something about this phrase seems off. I think "not constrained using plural binding" avoids anything too strange with mixed tense. Even "not plurally bound" feels weird.

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -300,34 +301,144 @@
+{% endhighlight %}
+
+Splats will iterate over the collection of selected elements and position elements
+next to each other. <a href="http://codepen.io/cbchouinard/pen/ogerrX" target="_blank">Live example.</a>
```
Same think in the linked codepen about article elements not being bound. I thought that might be wrong.

cbchouinard replied:
I removed them from the CodePen.

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -300,34 +301,144 @@
+<style type="text/gss">
+
+/* elements are not 'pluraly' binded */
+@h |-(section)-10-...-(article)-20-...-(div)...-| in(::window);
```
We seem to dive right into this complicated example with splats and then rush past them.

It would be good to build up to it, starting with something like this:

```
<section></section>
<section></section>
<section></section>

<style type="text/gss">
  @h |(section)...| in(::window);
</style>
```

Then, introducing gaps:

```
<section></section>
<section></section>
<section></section>

<style type="text/gss">
  @h |-(section)-...-| gap(10) outer-gap(20) in(::window);
</style>
```

After that, multiple elements and gaps wouldn't be so much of a leap.

paulyoung replied:
Same concerns as before about using hyphen for (outer) gap without using the (outer) gap function.

cbchouinard replied:
When this is fixed I'll adjust the doc: https://github.com/gss/vfl-compiler/issues/16

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -300,34 +301,144 @@
 
+In the previous example, the two `section` will be positioned next to each other with a gap of 10px. The
+three articles will be positioned next to each other with a gap of 20px. Finally the two div will be right next to each
+other without any gap.
```
There's some inconsistency with singular/plural here and how elements are being referenced.

We could either say:
* 1
  - two `section`s
  - three `article`s
  - two `div`s
* 2
  - two sections
  - three articles
  - two divs
* 3
  - two `section` elements
  - three `article` elements
  - two `div` elements

I'd probably go with option 2 or 3.

paulyoung replied:
Also, no mention of outer gap here even though it's in the VFL statement.

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -300,34 +301,144 @@
+other without any gap.
+
+<a name="virtuals"></a>
+## Virtual with VFL
```
I think we should refer to these as "Virtuals" or "Virtual elements".

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -300,34 +301,144 @@
+
+<a name="virtuals"></a>
+## Virtual with VFL
+[Virtuals](/guides/constraints) can be created and/or positionned using VFL expressions.
```
Should link to /guides/constraints/#virtuals

paulyoung replied:
positionned -> positioned

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -300,34 +301,144 @@
+## Virtual with VFL
+[Virtuals](/guides/constraints) can be created and/or positionned using VFL expressions.
 
+You can position virtual as any other elements :
```
Could we reword to:

>You can position a virtual the same as any other element:

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -300,34 +301,144 @@
+{% highlight css %}
 
-Chains are a powerful mechanism to take each element within an VFL alignment (excluding a containment element & points) and construct a chained constraint statement against a property of each element.  
+/* horizontally align two virtuals with a gap of 10px between each
```
No closing `*/`

paulyoung replied:
Maybe say "between _them_" since there are only two

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -300,34 +301,144 @@
 
-Chains are a powerful mechanism to take each element within an VFL alignment (excluding a containment element & points) and construct a chained constraint statement against a property of each element.  
+/* horizontally align two virtuals with a gap of 10px between each
+@h |-("area1")-10-("area2")-| in(::window)
```
No trailing semicolon

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -300,34 +301,144 @@
+<a name="rulesets"></a>
+## VFL rulesets
+
+You can use a VFL expression as the rule of a [rulesets](/guides/constraints/#rulesets). It will
```
"as the rule _in_ a _ruleset_ _and_ it will behave"

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -300,34 +301,144 @@
+
+You can use a VFL expression as the rule of a [rulesets](/guides/constraints/#rulesets). It will
+behave in the same way as a selector rule. GSS will create a collection of all elements involved
+in the VFL expression and iterates over them.
```
"and _iterate_ over them"

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -300,34 +301,144 @@
+in the VFL expression and iterates over them.
+
+This allows you to do interesing things. For example, if you want all the elements of the VFL expression
+to have the same width you can do the following:
```
Could we reword to something like this?

> This makes it possible to do interesting things. For example, to make all elements in the VFL expression have the same width:

(Fixes typo, removes "you")

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -300,34 +301,144 @@
+{% highlight css %}
+
+@h |-(section)-10-...-(article)-20-...-(div)...-| in(::window) {
+      width: == :next[width];
```
Not sure we should be promoting the use of `:next` here for a few reasons:
* couples the layout to the DOM
* the next element in the DOM may not be the next element in the VFL statement
* the same result can be achieved via something like `width: == [item-width];`

This is based on my understanding that `:next` means "the next sibling element in the DOM".

I'm not yet aware of any special handling around `:next` that addresses these concerns but would like to see a section in the docs that explains what happens in these scenarios.

cbchouinard replied:
We have a section in the selectors guide : Selectors and the DOM that explains the risk of using this.

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -300,34 +301,144 @@
+within your VFL expression.
+
+<a name="vflchains"></a>
+## VFL Chains
```
I thought these had been deprecated?

cbchouinard replied:
It would make sens to deprecate them since ruleset are much simpler. I'll confirm and remove the section if needed.

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -300,34 +301,144 @@
+{% endhighlight %}
+
+You can do with rulesets the same things as VFL Chains but they offer an alternative way to chains elements
+within your VFL expression.
```
This sentence seems as though it should be part of the section on chains since they haven't been introduced yet.

paulyoung replied:
This basically says "VFL chains allow you to chain elements", but what does that mean?

I'm struggling to come up with a definition, and I've used chains :trollface:

paulyoung replied:
Maybe we could say something like:

> VFL chains provide the same functionality as rulesets, but offer an alternative way to constrain related elements within a VFL expression.

paulyoung commented on guides/_posts/2014-01-5-vfl.html:
```diff
@@ -300,34 +301,144 @@
+
+Chains are a powerful mechanism to take each element within an VFL alignment (excluding a containment element & points)
+and construct a chained constraint statement against a property of each element.
+<a href="http://codepen.io/cbchouinard/pen/vEJqbG" target="_blank">Live example.</a>
```
This might be moot if chains are in fact deprecated, but if not, by saying that they're a "powerful mechanism" this sounds like we're promoting chains over rulesets.

The day rulesets came along I stopped using chains. I think we should consider removing this section altogether or at least putting it before the section on rulesets, and then introducing rulesets as the new, better way to do things.

paulyoung replied:
Suggesting that if this section sticks around that we reword to something like:

> Chains are a mechanism for constraining properties across multiple elements to a single value.

Then let the examples speak for themselves.

cbchouinard replied:
Just confirm with Dan, we'll remove "Chains" from the doc

cbchouinard replied:
I renamed constraints to its original name.

paulyoung commented:
What was the reasoning behind moving `/guides/ccss` to `/guides/constraints`? I think it would be better to keep the same URL to maintain any deep links to our site and for SEO purposes.

I also can't tell from the diff what was added or removed because it's considered to be a new file.

paulyoung commented:
That said, I'll hold off on reviewing that page until I hear back since it should be considerably easier to view the diff if the name is changed back.

paulyoung commented:
Disregard that, I can diff locally.

paulyoung commented on guides/_posts/2015-01-06-constraints.html:
```diff
@@ -0,0 +1,839 @@
+<a name="basicconstraint"></a>
+## Constraints, the basics
+
+Constraints express [relationships between variables](#constraintsaretwoways) that [may or may not hold](#mayofmaynothold). You can constrain any
```
I think this should be `twoway` (singular, not plural)

paulyoung replied:
`#mayofmaynothold` -> `#mayormaynothold`

paulyoung commented on guides/_posts/2015-01-06-constraints.html:
```diff
@@ -0,0 +1,839 @@
+## Constraints, the basics
+
+Constraints express [relationships between variables](#constraintsaretwoways) that [may or may not hold](#mayofmaynothold). You can constrain any
+numerical style property of an element, not just its position & size.
```
Could we change this to:

> Any numeric property of an element can be constrained, not just position & size.

paulyoung commented on guides/_posts/2015-01-06-constraints.html:
```diff
@@ -0,0 +1,839 @@
+Constraints express [relationships between variables](#constraintsaretwoways) that [may or may not hold](#mayofmaynothold). You can constrain any
+numerical style property of an element, not just its position & size.
+
+For example, if I want all paragraph tags to have line-height greater than 16px and less than 1/12th the size of the window:
```
In the same vein of removing "you", I think we should remove "I":

> For example, to specify that all paragraph tags have a line-height greater

paulyoung commented on guides/_posts/2015-01-06-constraints.html:
```diff
@@ -0,0 +1,839 @@
+
+{% endhighlight %}
+
+### Constraints declaration
```
Constraints -> Constraint

paulyoung commented on guides/_posts/2015-01-06-constraints.html:
```diff
@@ -0,0 +1,839 @@
+
+### Constraints declaration
+
+Let's go through the syntax for declaring a constraint:
```
I think being less informal would be better, something like:

> The syntax for declaring a constraint is as follows:

paulyoung commented on guides/_posts/2015-01-06-constraints.html:
```diff
@@ -0,0 +1,839 @@
+`p[line-height] >= 16;`
+
++ `p` is the selector. This constraint will apply to every `p` tag in the page.
++ `[]` is the style property selector.
```
The use of "style" here sounds confusing to me. Saying it's "the property selector" might be more accurate anyway.

paulyoung commented on guides/_posts/2015-01-06-constraints.html:
```diff
@@ -0,0 +1,839 @@
+
++ `p` is the selector. This constraint will apply to every `p` tag in the page.
++ `[]` is the style property selector.
++ `line-height` is the style property for which a value will be calculated by GSS.
```
Same thing here

paulyoung commented on guides/_posts/2015-01-06-constraints.html:
```diff
@@ -0,0 +1,839 @@
++ `p` is the selector. This constraint will apply to every `p` tag in the page.
++ `[]` is the style property selector.
++ `line-height` is the style property for which a value will be calculated by GSS.
++ `>=` defines an inequality constraint. You can declare inequality and equality constraint.
```
> You can declare inequality and equality constraint.

This seems redundant and sounds too informal to me.

paulyoung commented on guides/_posts/2015-01-06-constraints.html:
```diff
@@ -0,0 +1,839 @@
++ `[]` is the style property selector.
++ `line-height` is the style property for which a value will be calculated by GSS.
++ `>=` defines an inequality constraint. You can declare inequality and equality constraint.
++ `16` is the numerical value for the constraint in pixel unit. The default unit in GSS are pixels.
```
Something seems off here. Maybe it should be:

> `16` is the numerical value for the constraint in pixels (the default unit of measurement in GSS).

or something that doesn't mix singular and plural for "pixel", and "unit"

paulyoung commented on guides/_posts/2015-01-06-constraints.html:
```diff
@@ -0,0 +1,839 @@
+
+<a href="http://codepen.io/cbchouinard/pen/OPmdVL" target="_blank">Live example</a>
+
+<a name="constraintsaretwoways"></a>
```
`#constraintsaretwoways` -> `#constraintsaretwoway`

paulyoung commented on guides/_posts/2015-01-06-constraints.html:
```diff
@@ -0,0 +1,839 @@
+<a href="http://codepen.io/cbchouinard/pen/OPmdVL" target="_blank">Live example</a>
+
+<a name="constraintsaretwoways"></a>
+### Constraints are 2-ways
```
2-ways - > 2-way

paulyoung replied:
Actually, "two-way" is more consistent with other parts of the docs.

paulyoung commented on guides/_posts/2015-01-06-constraints.html:
```diff
@@ -0,0 +1,839 @@
+
+<a name="constraintsaretwoways"></a>
+### Constraints are 2-ways
+Do not confuse an equality constraint with your everyday variable assignment. For example, in a vanilla programming paradigm:
```
> Equality constraints should not be confused with everyday variable assignments.

(Not so informal)

paulyoung commented on guides/_posts/2015-01-06-constraints.html:
```diff
@@ -0,0 +1,839 @@
+
+Constraints are incrementally added to a `solver`, then the solver computes a feasible & optimal solution to all the constraints.
+With constraint programming you input the *what* you want to happen, not *how* it will happen. With procedural programming,
+you focus on the *how*.  This makes constraints a perfect fit for empowering declarative languages like CSS.
```
> Constraint programming takes *what* should happen as input, not *how* it will happen. Procedural programming focuses on the *how*.

(removes "you")

paulyoung commented on guides/_posts/2015-01-06-constraints.html:
```diff
@@ -0,0 +1,839 @@
+you focus on the *how*.  This makes constraints a perfect fit for empowering declarative languages like CSS.
+
+
+<a name="mayofmaynothold"></a>
```
mayofmaynothold -> mayormaynothold

paulyoung commented:
More to add, tomorrow :sleeping:

paulyoung commented:
I have more detailed feedback but in case I don't get to it soon (baby due to arrive soon), here are some more general questions/concerns:

* Renaming /guides/ccss to /guides/constraints seems unnecessary and could have adverse effects for SEO
* I think the sets of links at the top of each page could look better or be removed
* Maybe we should update the "GSS would not be possible without" section
* Why do we need `bower_components`?
* Should the codepen examples be under a personal account?
* No explanation of optional brackets in variables.

cbchouinard commented:
Paul, here are the answers to you questions:
- I renamed contraints to ccss 
- The links are temporary, Leigh will provide us a new design this week for a small redesign of the site
- We'll update the picture of the "GSS would not be possible..." with the pictures of everyone in the GSS team
- bower_component was there as a vestige :) I add it to gitignore 
- Good point for codepen, do we have a gss@thegrid.io or something? I'll ask around here what's the best way to do it. 
- I added a small mention in the variable section for optional brackets.
